src/native/clr/include/host/gc-bridge.hh

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <jni.h>
+#include <thread>
+#include <semaphore>
+#include <shared_mutex>
+#include <unordered_map>
+
+#include <shared/cpp-util.hh>
+
+struct JniObjectReferenceControlBlock
+{
+	jobject handle;
+	int handle_type;
+	jobject weak_handle;
+	int refs_added;
+};
+
+struct HandleContext
+{
+	JniObjectReferenceControlBlock* control_block;
+};
+
+struct StronglyConnectedComponent
+{
+	size_t Count;
+	HandleContext** Contexts;
+};
+
+struct ComponentCrossReference
+{
+	size_t SourceGroupIndex;
+	size_t DestinationGroupIndex;
+};
+
+struct MarkCrossReferencesArgs
+{
+    size_t ComponentCount;
+    StronglyConnectedComponent* Components;
+    size_t CrossReferenceCount;
+    ComponentCrossReference* CrossReferences;
+};
+
+using BridgeProcessingStartedFtn = void (*)();
+using BridgeProcessingFinishedFtn = void (*)(MarkCrossReferencesArgs*);
+using BridgeProcessingFtn = void (*)(MarkCrossReferencesArgs*);
+
+namespace xamarin::android {
+	class GCBridge
+	{
+	public:
+		static void wait_for_bridge_processing () noexcept;
```

Comment:
This is a simple enough method, might want to make it inline here.

Reply:
It is true that it could be inlined here but I prefer having all the places which touch the mutex+semaphore to be next to each other in the `.cc` file. If you're ok with it, I would keep this part of the code as is.